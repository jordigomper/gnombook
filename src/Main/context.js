import React, { createContext, useReducer, useEffect } from "react";
import { isArray } from "util";

const STATE_INITIAL_VALUE = {
  habitants: [],
  professions: []
};

const APIContext = createContext([]);

const APIProvider = ({ children }) => {
  function reducer(state, action) {
    const currentState = JSON.parse(JSON.stringify(state));

    switch (action.type) {
      case "fetch":
        const professionsList = currentState.professions;
        action.data.map(({ professions }) =>
          professions.map(profession => {
            const current = profession.trim().toLowerCase();
            !professionsList.includes(current) && professionsList.push(current);
          })
        );
        return { habitants: action.data, professions: professionsList };
      default:
        return state;
    }
  }

  const [state, dispatch] = useReducer(reducer, STATE_INITIAL_VALUE);

  async function fetchData() {
    const localData = localStorage.getItem("habitantsLocal");
    localData && dispatch({ type: "fetch", data: JSON.parse(localData) });

    const response = await fetch(process.env.REACT_APP_API_URL)
      .then(response => response.json())
      .then(({ Brastlewark }) => Brastlewark)
      .catch(error => console.error(error.message));

    if (response && isArray(response)) {
      console.log(response);
      dispatch({ type: "fetch", data: response });
      localStorage.setItem("habitantsLocal", JSON.stringify(response));
    }
  }

  useEffect(() => {
    fetchData();
  }, []);

  return <APIContext.Provider value={state}>{children}</APIContext.Provider>;
};

export { APIContext, APIProvider };
