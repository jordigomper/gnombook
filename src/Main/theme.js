export const theme = {
  breakPoints: {
    phone: "991px",
    tablet: "768px",
    desktop: "575px"
  }
};
