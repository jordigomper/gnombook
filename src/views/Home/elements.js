import styled from "@emotion/styled";

export const List = styled.div`
  display: grid;
  justify-content: center;
  grid-template-columns: 40% 40%;
  grid-column-gap: 8%;
  -webkit-transition: heigth 12s;
  transition: heigth 12s;

  @media (min-width: 950px) {
    margin: 30px 0 60px 0;
    grid-template-columns: 20% 20% 20%;
  }
`;

export const PaginatorPanel = styled.div`
  margin: 10px 5%;
  padding: 0 5%;
  flex: 1;
  display: flex;
  justify-content: space-between;
  align-items: center;
  align-items: flex-end;
  p {
    margin: 0 0 11px 0;
  }
`;

export const Icon = styled.img`
  cursor: pointer;
  width: 40px;
  background: #4267b273;
  border-radius: 100%;
  ${({ disabled }) =>
    disabled &&
    `
    cursor: initial;
    opacity: 0;
  `};
`;
