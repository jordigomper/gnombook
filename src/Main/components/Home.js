import React, { useEffect, useContext, useState } from "react";
import { hasSome } from "../../module/utils";
import { usePaginator } from "../../module/hooks";
import { APIContext } from "../context";
import Searchbar from "./Searchbar";
import Card from "./Card";
import FilterPanel from "./FilterPanel";
import styled from "@emotion/styled";

const arrow_r = require("../../assets/icons/chevron_right.svg");
const arrow_l = require("../../assets/icons/chevron_left.svg");

// @media only screen and (min-device-width : 768px) and (max-device-width : 1024px) {
//   /* Styles */
//   }
// @media only screen and (min-device-width : 768px) and (max-device-width : 1024px) and (orientation : landscape) and (-webkit-min-device-pixel-ratio : 2) {
//   /* Styles */
//   }

const SKIP_PAGINATION = 6;

const List = styled.div`
  display: grid;
  justify-content: center;
  grid-template-columns: 40% 40%;
  grid-column-gap: 8%;
  -webkit-transition: heigth 12s;
  transition: heigth 12s;
`;

const PaginatorPanel = styled.div`
  margin: 10px 5%;
  padding: 0 5%;
  flex: 1;
  display: flex;
  justify-content: space-between;
  align-items: center;
  align-items: flex-end;
  p {
    margin: 0 0 11px 0;
  }
`;

const Icon = styled.img`
  width: 40px;
  background: gray;
  border-radius: 100%;
`;

const Home = () => {
  const {
    page,
    setPage,
    nextPage,
    prevPage,
    currentPage,
    totalPages
  } = usePaginator();

  const { habitants, professions } = useContext(APIContext);
  const [habitantsCookedData, setHabitantsCookedData] = useState(habitants);

  const [searchTerm, setSearchTerm] = useState("");
  const [filter, setFilter] = useState([]);

  window.f = () => console.log(filter);

  const toggleFilter = element => {
    const hasIndexArray = filter.indexOf(element);
    const newFilterState = filter;

    // TODO don't working
    hasIndexArray > -1
      ? newFilterState.slice(hasIndexArray, 1)
      : newFilterState.push(element);

    setFilter([...newFilterState]);
  };

  useEffect(() => {
    let cookedData = habitants;
    if (searchTerm.length > 0) {
      cookedData = cookedData.filter(({ name }) =>
        name
          .trim()
          .toLowerCase()
          .includes(searchTerm.trim().toLowerCase())
      );
    }

    if (filter.length > 0) {
      cookedData = cookedData.filter(({ professions }) =>
        hasSome(filter, professions)
      );
    }
    setPage(0);
    setHabitantsCookedData(cookedData);
  }, [habitants, searchTerm, filter]);

  window.pa = page => setPage(page * SKIP_PAGINATION);

  return (
    <>
      <Searchbar value={searchTerm} onChange={setSearchTerm} />

      <FilterPanel buttons={professions} onClick={toggleFilter} />

      <List>
        {habitantsCookedData
          .slice(page, page + SKIP_PAGINATION)
          .map(habitant => (
            <Card key={habitant.id} {...habitant} />
          ))}
      </List>

      <PaginatorPanel>
        <Icon
          onClick={currentPage !== 1 && prevPage}
          src={arrow_l}
          alt={"previous"}
        />
        <p>
          <b>
            {currentPage} / {totalPages(habitantsCookedData)}
          </b>
        </p>
      </PaginatorPanel>
    </>
  );
};

export default Home;
