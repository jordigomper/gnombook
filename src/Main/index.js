import React from "react";
import styled from "@emotion/styled";
import { APIProvider } from "./context";
import Home from "./components/Home";

const Header = styled.h1`
  width: 100%;
  margin: 0;
  background-color: blue;
  text-align: center;
`;

const Main = styled.div`
  display: flex;
  flex-direction: column;
  justifi-content: center;
  align-self: center;
`;

function App() {
  return (
    <Main>
      <APIProvider>
        <Header>nice to meet you!</Header>
        <Home />
      </APIProvider>
    </Main>
  );
}

export default App;
