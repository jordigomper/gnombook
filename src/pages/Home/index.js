import React, { useEffect, useState } from "react";
import { isString, isNumber } from "util";

import { Searchbar, Card, FilterPanel } from "../../components";
import { List, PaginatorPanel, Icon } from "./elements";
import { hasSome } from "../../module/utils";
import { usePaginator } from "../../module/hooks";
import { useAPIContext } from "../../App/context";

const arrow_r = require("../../assets/icons/chevron_right.svg");
const arrow_l = require("../../assets/icons/chevron_left.svg");

export const Home = () => {
  const {
    page,
    setPage,
    nextPage,
    prevPage,
    currentPage,
    totalPages,
    itemsForPage,
  } = usePaginator();

  const { habitants, professions } = useAPIContext();
  const [habitantsCookedData, setHabitantsCookedData] = useState([]);

  const [searchTerm, setSearchTerm] = useState("");
  const [filter, setFilter] = useState([]);

  // add or remove the selector pushed
  function toggleFilter(element) {
    const hasIndexArray = filter.indexOf(element);
    const newFilterState = JSON.parse(JSON.stringify(filter));

    hasIndexArray > -1
      ? newFilterState.splice(hasIndexArray, 1)
      : newFilterState.push(element);

    setFilter([...newFilterState]);
  }

  useEffect(() => {
    const cookedData = filterData;
    setPage(0);
    setHabitantsCookedData(cookedData);
  }, [habitants, searchTerm, filter]);

  function filterData() {
    let cookedData = habitants;

    // check id format
    cookedData = cookedData.filter(
      (hab) =>
        hab.hasOwnProperty("id") && (isNumber(hab.id) || isString(hab.id))
    );

    // filter searchbar
    if (searchTerm.length > 0) {
      cookedData = cookedData.filter(({ name }) =>
        name.trim().toLowerCase().includes(searchTerm.trim().toLowerCase())
      );
    }

    // filter selectors
    if (filter.length > 0) {
      cookedData = cookedData.filter(({ professions }) =>
        hasSome(filter, professions)
      );
    }
    return cookedData;
  }

  return (
    <>
      <Searchbar
        className="searchbar"
        value={searchTerm}
        onChange={setSearchTerm}
      />

      <FilterPanel buttons={professions} onClick={toggleFilter} />

      <List className="list">
        {habitantsCookedData
          .slice(page, page + itemsForPage)
          .map((habitant) => (
            <Card className="list__item" key={habitant.id} {...habitant} />
          ))}
      </List>

      <PaginatorPanel>
        <Icon
          disabled={currentPage === 1}
          onClick={() => prevPage(currentPage === 1)}
          src={arrow_l}
          alt={"previous"}
        />
        <p>
          <b>
            {currentPage} / {totalPages(habitantsCookedData)}
          </b>
        </p>
        <Icon
          disabled={currentPage >= totalPages(habitantsCookedData)}
          onClick={() =>
            nextPage(currentPage >= totalPages(habitantsCookedData))
          }
          src={arrow_r}
          alt={"next"}
        />
      </PaginatorPanel>
    </>
  );
};
