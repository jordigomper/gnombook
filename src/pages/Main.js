import React from "react";
import { Router } from "@reach/router";
import { ThemeProvider } from "emotion-theming";
import styled from "@emotion/styled";

// import Header from "./layouts/Header";
import { Header } from "./layouts";
import Home from "./Home";
import Profile from "./Profile";

import { APIProvider } from "../context";

import { theme } from "../theme.js";

const Main = styled.div`
  min-height: 100vh;
  background-color: #e9ebee;
  display: flex;
  flex-direction: column;
  align-self: center;
`;

const RouterStyled = styled(Router)`
  background-color: white;
  flex: 1;
  display: flex;
  flex-direction: column;
  ${({ theme: { breakPoints } }) => `
    @media (min-width: ${breakPoints.tablet}px) {
      margin: 0 15%;
      padding: 0 15px;
      border: 1px solid #dddfe2;
      border-radius: 3px;
    }
  `}
`;

function App() {
  return (
    <Main>
      <ThemeProvider theme={theme}>
        <Header>nice to meet you!</Header>
        <APIProvider>
          <RouterStyled>
            <Home exact path="/" />
            <Profile exact path="/profile/:id" />
          </RouterStyled>
        </APIProvider>
      </ThemeProvider>
    </Main>
  );
}

export default App;
