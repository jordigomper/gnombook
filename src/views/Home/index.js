import React, { useEffect, useState } from "react";
import { isString, isNumber } from "util";

import { Searchbar, Card, FilterPanel } from "../../components";
import { List, PaginatorPanel, Icon } from "./elements";
import { hasSome } from "../../module/utils";
import { usePaginator } from "../../module/hooks";
import { useAPIContext } from "../../App/context";

const rightArrow = require("../../assets/icons/chevron_right.svg");
const leftArrow = require("../../assets/icons/chevron_left.svg");

export const Home = () => {
  const {
    page,
    setPage,
    nextPage,
    prevPage,
    currentPage,
    totalPages,
    itemsForPage,
  } = usePaginator();

  const { habitants, professions } = useAPIContext();
  const [habitantsCookedData, setHabitantsCookedData] = useState([]);

  const [searchTerm, setSearchTerm] = useState("");
  const [filter, setFilter] = useState([]);

  // add or remove the selector pushed
  function toggleFilter(element) {
    const hasIndexArray = filter.indexOf(element);
    const newFilterState = JSON.parse(JSON.stringify(filter));

    hasIndexArray > -1
      ? newFilterState.splice(hasIndexArray, 1)
      : newFilterState.push(element);

    setFilter([...newFilterState]);
  }

  useEffect(() => {
    const cookedData = filterData;
    setPage(0);
    setHabitantsCookedData(cookedData);
  }, [habitants, searchTerm, filter]);

  function filterData() {
    let cookedData = habitants;

    // check id format
    cookedData = cookedData.filter(
      (hab) =>
        hab.hasOwnProperty("id") && (isNumber(hab.id) || isString(hab.id))
    );

    // filter searchbar
    if (searchTerm.length > 0) {
      cookedData = cookedData.filter(({ name }) =>
        name.trim().toLowerCase().includes(searchTerm.trim().toLowerCase())
      );
    }

    // filter selectors
    if (filter.length > 0) {
      cookedData = cookedData.filter(({ professions }) =>
        hasSome(filter, professions)
      );
    }
    return cookedData;
  }

  return (
    <>
      <Searchbar
        className="searchbar"
        value={searchTerm}
        onChange={setSearchTerm}
      />

      <FilterPanel buttons={professions} onClick={toggleFilter} />

      <List className="list">
        {habitantsCookedData
          .slice(page, page + itemsForPage)
          .map((habitant) => (
            <Card className="list__item" key={habitant.id} {...habitant} />
          ))}
      </List>

      <PaginatorPanel>
        <Icon
          disabled={currentPage === 1}
          onClick={() => prevPage(currentPage === 1)}
          src={leftArrow}
          alt={"previous"}
        />
        <p>
          <b>
            {currentPage} / {totalPages(habitantsCookedData)}
          </b>
        </p>
        <Icon
          disabled={currentPage >= totalPages(habitantsCookedData)}
          onClick={() =>
            nextPage(currentPage >= totalPages(habitantsCookedData))
          }
          src={rightArrow}
          alt={"next"}
        />
      </PaginatorPanel>
    </>
  );
};
