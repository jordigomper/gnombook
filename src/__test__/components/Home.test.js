import React from "react";
import { mount, shallow } from "enzyme";
import Home from "../../Main/components/Home";
import * as AppContext from "../../Main/context";

const gnome = {
  age: 306,
  friends: ["Cogwitz Chillwidget", "Tinadette Chillbuster"],
  hair_color: "Pink",
  height: 107.75835,
  id: 0,
  name: "Tobus Quickwhistle",
  professions: [
    "Metalworker",
    "Woodcarver",
    "Stonecarver",
    " Tinker",
    "Tailor",
    "Potter"
  ],
  thumbnail:
    "http://www.publicdomainpictures.net/pictures/10000/nahled/thinking-monkey-11282237747K8xB.jpg",
  weight: 39.065952
};

const gnomeBadFormat = {
  age: 306,
  friends: ["Cogwitz Chillwidget", "Tinadette Chillbuster"],
  hair_color: "Pink",
  height: 107.75835,
  name: "Tobus Quickwhistle",
  professions: [
    "Metalworker",
    "Woodcarver",
    "Stonecarver",
    " Tinker",
    "Tailor",
    "Potter"
  ],
  thumbnail:
    "http://www.publicdomainpictures.net/pictures/10000/nahled/thinking-monkey-11282237747K8xB.jpg",
  weight: 39.065952
};

describe("Home component", () => {
  it("renders correctly", () => {
    const wrapper = shallow(<Home />);

    expect(wrapper).toMatchSnapshot();
  });

  it("renders list", () => {
    jest.spyOn(AppContext, "useAPIContext").mockImplementation(() => {
      return { habitants: [gnome, gnomeBadFormat] };
    });

    const wrapper = mount(<Home />);

    console.log(wrapper.debug());

    expect(wrapper.find(".list .list__item")).toHaveLength(1);
  });
});
