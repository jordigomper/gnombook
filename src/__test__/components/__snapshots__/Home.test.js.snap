// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component renders correctly 1`] = `ShallowWrapper {}`;

exports[`Home component renders correctly 2`] = `ShallowWrapper {}`;

exports[`Home component starts with a count of 0 1`] = `ShallowWrapper {}`;
